File: Sources/SwiftMTH/CommFlags.swift
/// Communication flags for a telnet session.
///
/// Tracks negotiated protocol features and connection state.
public struct CommFlags: OptionSet, Sendable {
    public let rawValue: Int

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    public static let disconnect  = CommFlags(rawValue: 1 << 0)
    public static let password    = CommFlags(rawValue: 1 << 1)
    public static let remoteEcho  = CommFlags(rawValue: 1 << 2)
    public static let eor         = CommFlags(rawValue: 1 << 3)
    public static let msdpUpdate  = CommFlags(rawValue: 1 << 4)
    public static let colors256   = CommFlags(rawValue: 1 << 5)
    public static let utf8        = CommFlags(rawValue: 1 << 6)
    public static let gmcp        = CommFlags(rawValue: 1 << 7)
}

File: kotlin/mth-core/src/main/kotlin/mth/core/CommFlags.kt
package mth.core

@JvmInline
value class CommFlags(val rawValue: Int = 0) {
    operator fun contains(flag: CommFlags): Boolean = rawValue and flag.rawValue == flag.rawValue
    fun insert(flag: CommFlags) = CommFlags(rawValue or flag.rawValue)
    fun remove(flag: CommFlags) = CommFlags(rawValue and flag.rawValue.inv())
    fun isEmpty(): Boolean = rawValue == 0

    companion object {
        val DISCONNECT = CommFlags(1 shl 0)
        val PASSWORD = CommFlags(1 shl 1)
        val REMOTE_ECHO = CommFlags(1 shl 2)
        val EOR = CommFlags(1 shl 3)
        val MSDP_UPDATE = CommFlags(1 shl 4)
        val COLORS_256 = CommFlags(1 shl 5)
        val UTF8 = CommFlags(1 shl 6)
        val GMCP = CommFlags(1 shl 7)
    }
}
